enterprise_gateway/client/gateway_client.py

```diff
@@ -139,20 +132,29 @@ def shutdown(self):
         # Terminate thread, close socket and clear queues.
         self.shutting_down = True
 
-        if self.kernel_socket:
+        if hasattr(self, 'kernel_socket') and self.kernel_socket:
```

Comment:
Are these attributes checks necessary?  How can `KernelClient.shutdown()` be called if there isn't a `KernelClient` instance - which is the only way these attributes would not be present?

Reply:
I believe they are necessary, since when I tested exceptions on the `websocket.create_connection()` in `KernelClient.__init__()`, this line would crash as `kernel_socket` was not defined. The same occurred for `response_queues` and `response_reader`.